### PR TITLE
FINAL CAPSTONE: Hotel Reservation Application

### PR DESCRIPTION
**Pull request by carlylechia**
## App Description
This application, we christened, 'Hove' which stands for Hotel ReserVe, lets a user do just that, reserve hotels. Admins have the ability to add, update and delete hotels and their information to the database, and a user can reserve any hotel available, according to their needs.

## Project Requirement
-     Create an API with rails, using PostgreSQL as a database, ERD is found in the Readme.MD file.
-     Test all API endpoints with Rswag and generate the API documentation.
-     Build front-end UI with React and Redux.
-     Style the app and make it responsive, following the design by [Murat Korkmaz on Behance](https://www.behance.net/muratk) 

## Team.
Our team started off as a team of five, but towards the end of the first week, one member, @Pazzo97, unfortunately, could not continue working with us for personal reasons.
Therefore, the surviving team consists of four members:
* @Lembani 
* @frank1738 
* @ekenecf 
* @carlylechia 

Each member contributed equally, to the realization of this project.

## Describe your changes

-  In this project, our team did our best to implement all the project requirements.
- The subject of our app was Hotel Reservations.
- For the backend, we built a RESTFUL API using Ruby on Rails, and a PostgreSQL database.
- The API documentation can be accessed through [this endpoint](https://stark-badlands-38572.herokuapp.com/api-docs)
- Our backend was deployed on Heroku and can be accessed via [this domain](https://stark-badlands-38572.herokuapp.com/api/v1/hotels)
- The front end was built with React and Redux.
- It was deployed to Netlify and can be accessed by clicking [this link](https://guileless-sprinkles-294f6c.netlify.app).

### What kind of change does this PR introduce?
Dev.

### In which branch do you want the changes to merge?

Main.

### Documentation updated ?
Yes

### Checklist

- [ ] No ES Lint Error
- [ ] No Style Lint Error
- [ ] No Rubocop Error
- [ ] All tests and specs passing.

## KanBan Board Usage

This project used a kanban board to distribute tasks among teammates. Here's [a link](https://github.com/Lembani/hotel-reservation-backend/projects/1) to the board as it is currently.
Screenshots of the initial and current states of the board are also provided below.

### Initial state
image

### Current (Final) State
kanban-final-state

## Front-end PR:
This is [a link to the PR](https://github.com/Lembani/hotel-reservation-frontend/pull/18) in the front-end repository

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,123 @@
...
+- Ruby version 3.1.2
+- Rails (RoR) version 7.0.4
+- Postgresql
+- Rubocop
+
+## Front-End
+
+You can access the front-end of this application via [this link](https://github.com/Lembani/hotel-reservation-frontend.git).
+
+## SetUp
```
Comment (Meltrust) [inline README.md:20]: - [ ] On your readme, please include the commands that are missing to run the project:`rails db:create`, `rails db:migrate` and `rails db:seed`.

That way, your users will be able to run the project 💪

File: Gemfile [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+gem 'rubocop', '>= 1.0', '< 2.0'
+
+gem 'rack-cors'
+
+gem 'devise'
+
+gem 'devise-jwt'
+
+gem 'cancancan'
```
Comment (Meltrust) [inline Gemfile:24]: - [ ] It seems that almost all the code related to `cancancan` has been either removed or commented out.


If not using `cancancan` anymore, please consider deleting the gem and all the related code from the corresponding files (like ability.rb).

File: app/controllers/api/v1/hotels_controller.rb [hunk 3]
```diff
@@ -0,0 +1,85 @@
+class Api::V1::HotelsController < ApplicationController
+  # before_action :logged_in, except: %i[index show]
+
+  # before_action :user_ability, except: %i[index show]
```
Comment (Meltrust) [inline app/controllers/api/v1/hotels_controller.rb:4]: - [ ] Please, remove these old pieces of commented code to improve the cleanliness of the repo 💪

File: app/controllers/api/v1/hotels_controller.rb [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    params.require(:hotel).permit(:name, :price, :description, :country, :city, :address, :image_url, :category_id)
+  end
+
+  # def user_ability
+  #   authorize! :manage, @hotel
+  # rescue CanCan::AccessDenied
+  #   render json: { errors: 'You are not authorized to perform this action' },
+
+  #          status: :unauthorized
+  # end
```
Comment (Meltrust) [inline app/controllers/api/v1/hotels_controller.rb:84]: - [ ] Please, remove these old pieces of commented code to improve the cleanliness of the repo 💪

File: app/controllers/api/v1/reservations_controller.rb [hunk 5]
```diff
@@ -0,0 +1,64 @@
+class Api::V1::ReservationsController < ApplicationController
+  before_action :set_reservation_params, only: %i[show update destroy]
+  # before_action :logged_in, except: %i[index show]
+  # before_action :user_ability, except: %i[update destroy]
```
Comment (Meltrust) [inline app/controllers/api/v1/reservations_controller.rb:4]: - [ ] Please, remove these old pieces of commented code to improve the cleanliness of the repo 💪

File: app/controllers/api/v1/reservations_controller.rb [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  def reservation_params
+    params.require(:reservation).permit(:reason, :duration, :start_day, :end_day, :hotel_id, :user_id)
+  end
+
+  # def user_ability
+  #   authorize! :manage, @reservation
+  # rescue CanCan::AccessDenied
+  #   render json: { errors: 'You are not authorized to perform this action' },
+  #          status: :unauthorized
+  # end
```
Comment (Meltrust) [inline app/controllers/api/v1/reservations_controller.rb:63]: - [ ] Please, remove these old pieces of commented code to improve the cleanliness of the repo 💪

**Review by Meltrust — CHANGES_REQUESTED** [review]
Hi @carlylechia,

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation, but you are almost there!






### To highlight:

-  Tests are passing✔️
-  App is working ✔️ 
-  Good readme ✔️ 

You can do it

## Suggested Changes ♻️
_Check the comments under the review._

You can use as many of my suggestions as you want.  If there is anything you would like to skip - feel free to do that.  However, I strongly recommend you take them into account as they can make your code better.


Cheers and Happy coding!👏👏👏

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------

File: README.md (hunk 1 above)
Comment (Lembani) [reply README.md:20]: Resolved, thank you.

File: app/controllers/api/v1/hotels_controller.rb (hunk 3 above)
Comment (Lembani) [reply app/controllers/api/v1/hotels_controller.rb:4]: Resolved, thanks.

File: app/controllers/api/v1/hotels_controller.rb (hunk 4 above)
Comment (Lembani) [reply app/controllers/api/v1/hotels_controller.rb:84]: Resolved, thanks.

File: app/controllers/api/v1/reservations_controller.rb (hunk 5 above)
Comment (Lembani) [reply app/controllers/api/v1/reservations_controller.rb:4]: Resolved, thanks.

File: app/controllers/api/v1/reservations_controller.rb (hunk 6 above)
Comment (Lembani) [reply app/controllers/api/v1/reservations_controller.rb:63]: Issue resolved, thanks.

File: Gemfile (hunk 2 above)
Comment (Lembani) [reply Gemfile:24]: Issue resolved, thank you.

**Pull request merged**